lighthouse-core/lib/dependency-graph/simulator/simulator.js

```diff
@@ -134,6 +137,12 @@ class Simulator {
    * @param {number} queuedTime
    */
   _markNodeAsReadyToStart(node, queuedTime) {
+    const firstNodeIndexWithGreaterStartTime = this._cachedNodeListByStartTime
+      .findIndex(candidate => candidate.startTime > node.startTime);
+    const insertionIndex = firstNodeIndexWithGreaterStartTime === -1 ?
+      this._cachedNodeListByStartTime.length : firstNodeIndexWithGreaterStartTime;
+    this._cachedNodeListByStartTime.splice(insertionIndex, 0, node);
```

Comment:
a) How do we only end up with 2x test coverage of this? Or is codecov messed up? Something seems wrong with that.
b) `_nodes[NodeState.ReadyToStart]` (and the local `nodesReadyToStart`) appears unused now and I don't believe we use the uniqueness of `Set` for anything. Should we get rid of it? (I regret all the `_nodes[NodeState]` business now, incidentally :)
c) We only traverse in order of start time, should we switch out with something like a heap? Iterating over them in order goes up to n log n, but insertion/deletion drops to log n

Reply:
> this LGTM, but if you want to do a little more perf golfing... :)

well, actually we really should get rid of `_nodes[NodeState.ReadyToStart]` or at least leave a note to refactor since it's not used anymore.

Reply:
> Or is codecov messed up? 

codecov is definitely messed up, this will get hit for every node on every single simulation call which we have lots of coverage for

> _nodes[NodeState.ReadyToStart] (and the local nodesReadyToStart) appears unused now and I don't believe we use the uniqueness of Set for anything. Should we get rid of it?

seeing as I didn't really think it was necessary in the first place, I'm open to it 😉 

I will say the indexed by state has grown on me though, the clarity at least that they're moving through the system in these sets is appreciated when returning to the code after a long absence. I added a comment explaining this for now.

Reply:
> I will say the indexed by state has grown on me though, the clarity at least that they're moving through the system in these sets is appreciated when returning to the code after a long absence.

that's a fair point, but maybe there's a better way to handle enumerated state like that. As it is, it's difficult to find all the places we insert nodes :)

Reply:
> should we switch out with something like a heap?

I'm somewhat iffy of the ROI on that. Math-wise I'm with ya, but the fact that we just moved from a situation that Math-wise was `O(n log n)` iteration and `O(1)` insertion makes me skeptical it'll be worth the extra complexity since our current tradeoff favored lowering iteration time.